src/main/kotlin/org/rust/ide/inspections/RsDeprecationInspection.kt

```diff
@@ -0,0 +1,99 @@
+            if (ref is RsModDeclItem || ref !is RsWeakReferenceElement) return
+
+            var original = ref.reference?.resolve() ?: return
+
+            val identifier = ref.referenceNameElement ?: ref
+            original = when (original) {
+                is RsFile -> original.declaration
+                is RsAbstractable -> original.superItem
+                else -> original
+            } ?: original
+
+            if (original is RsOuterAttributeOwner && ref.parents().none { it.hasAllowDeprecatedAttribute() }) {
+                checkAndRegisterAsDeprecated(identifier, original, holder)
+            }
+        }
+    }
+
+    private fun checkAndRegisterAsDeprecated(identifier: PsiElement, original: RsOuterAttributeOwner, holder: ProblemsHolder) {
+        original.outerAttrList
+            .mapNotNull { DeprecatedAttribute.from(it) }
+            .firstOrNull()
+            ?.let {
+                holder.registerProblem(identifier, it.getMessage(identifier.text), ProblemHighlightType.LIKE_DEPRECATED)
+            }
+    }
+
+    private fun PsiElement.hasAllowDeprecatedAttribute(): Boolean = this is RsOuterAttributeOwner &&
+        outerAttrList
+            .any {
+                ALLOW_ATTR_NAME == it.metaItem.identifier?.text &&
+                    true == it.metaItem
+                    .metaItemArgs
+                    ?.metaItemList
+                    ?.any { param -> DEPRECATED_PARAM_NAME == param.identifier?.text }
+            }
+
+    companion object {
+        private const val ALLOW_ATTR_NAME: String = "allow"
+        private const val DEPRECATED_PARAM_NAME: String = "deprecated"
+    }
+}
+
+private class DeprecatedAttribute(val note: String?, val since: String?) {
+
+    fun getMessage(item: String): String = buildString {
+        append("'$item' is marked as deprecated")
+        if (since != null) append(" since version $since")
+        if (note != null) append(" ($note)")
+    }
+
+    companion object {
+        private const val DEPRECATED_ATTR_NAME: String = "deprecated"
+        private const val RUSTC_DEPRECATED_ATTR_NAME: String = "rustc_deprecated"
+
+        private const val SINCE_PARAM_NAME: String = "since"
+        private const val NOTE_PARAM_NAME: String = "note"
+        private const val REASON_PARAM_NAME: String = "reason"
+
+        fun from(attr: RsOuterAttr): DeprecatedAttribute? {
+            fun List<RsMetaItem>.getByName(name: String): String? =
+                firstOrNull { name == it.identifier?.text }?.litExpr?.stringLiteralValue
+
+            fun RsOuterAttr.extract(noteParamName: String, sinceParamName: String): DeprecatedAttribute {
+                val params = metaItem.metaItemArgs?.metaItemList
+                val note = params?.getByName(noteParamName)
+                val since = params?.getByName(sinceParamName)
+                return DeprecatedAttribute(note, since)
+            }
+
+            val identifier = attr.metaItem.identifier?.text
```

Comment:
Let's move all search-related code into `org.rust.lang.core.psi.ext.QueryAttributes`
Something like this:
```kotlin
fun findDeprecatedAttribute(): RsMetaItem? {
    // some code
}
```